packages/core/src/codewhispererChat/tools/grepSearch.ts

```diff
@@ -0,0 +1,214 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import * as vscode from 'vscode'
+import { getLogger } from '../../shared/logger/logger'
+import { sanitizePath, InvokeOutput, OutputKind } from './toolShared'
+import fs from '../../shared/fs/fs'
+import { Writable } from 'stream'
+import { ChildProcess, ChildProcessOptions } from '../../shared/utilities/processUtils'
+import { rgPath } from 'vscode-ripgrep'
+import path from 'path'
+
+export interface GrepSearchParams {
+    path?: string
+    query: string
+    caseSensitive?: boolean
+    excludePattern?: string
+    includePattern?: string
+    explanation?: string
+}
+
+export class GrepSearch {
+    private fsPath: string | undefined
+    private query: string
+    private caseSensitive: boolean
+    private excludePattern?: string
+    private includePattern?: string
+    private readonly logger = getLogger('grepSearch')
+
+    constructor(params: GrepSearchParams) {
+        this.fsPath = params.path
+        this.query = params.query
+        this.caseSensitive = params.caseSensitive ?? false
+        this.excludePattern = params.excludePattern
+        this.includePattern = params.includePattern
+    }
+
+    public async validate(): Promise<void> {
+        if (!this.query || this.query.trim().length === 0) {
+            throw new Error('Grep search query cannot be empty.')
+        }
+
+        // Handle optional path parameter
+        if (!this.fsPath || this.fsPath.trim().length === 0) {
+            // Use current workspace folder as default if path is not provided
+            const workspaceFolders = vscode.workspace.workspaceFolders
+            if (!workspaceFolders || workspaceFolders.length === 0) {
+                throw new Error('Path cannot be empty and no workspace folder is available.')
+            }
+            this.fsPath = workspaceFolders[0].uri.fsPath
+            this.logger.debug(`Using default workspace folder: ${this.fsPath}`)
+        }
+
+        const sanitized = sanitizePath(this.fsPath)
+        this.fsPath = sanitized
+
+        const pathUri = vscode.Uri.file(this.fsPath)
+        let pathExists: boolean
+        try {
+            pathExists = await fs.existsDir(pathUri)
+            if (!pathExists) {
+                throw new Error(`Path: "${this.fsPath}" does not exist or cannot be accessed.`)
+            }
+        } catch (err) {
+            throw new Error(`Path: "${this.fsPath}" does not exist or cannot be accessed. (${err})`)
+        }
+    }
+
+    public queueDescription(updates: Writable): void {
+        const searchDirectory = this.getSearchDirectory(this.fsPath)
```

Comment:
can we extract this outside, this is being called multiple times

Reply:
done